Step4 : 자동차게임 구현 (우승자)

baek0318 opened the pull request:
Step3에서 피드백주신 부분과
Step4에서 새롭게 추가된 스펙을 작성하였습니다
리뷰 부탁드리겠습니다!!

baek0318 committed:
step4 : move가 Int 받던것을 DiceResult를 받도록 변경
- 기존에 Int를 받음으로써 불명확했던 move api를 더 명확하게 어떤인자를 받을 수 있도록 변경

baek0318 committed:
step4 : Dice.pitch가 DiceResult를 반환하도록 변경

baek0318 committed:
step4 : RacingResult 객체를 별도로 생성함
- 기존에는 List<List<Int>>로 반환하던 값을 RacingGameResult객체를 반환하도록 변경
- 값을 감싸게 됨으로 외부 의존성들이 어떤 객체가 넘어오는지 정확하게 인지하게 됨

baek0318 committed:
step4 : RacingGameResult에 대해서 print하도록 변경

baek0318 committed:
step4 : Car가 이름을 가지도록 수정

baek0318 committed:
step4 : 이름에 대해서 입력을 받을 수 있도록 수정

baek0318 committed:
step4 : 우승자를 판별할 수 있도록 수정

baek0318 committed:
step4 : 우승자를 노출할 수 있도록 수정

baek0318 committed:
step4 : 함수의 depth가 깊지 않도록 수정

baek0318 committed:
step4 : 요구사항 수정

laco-dev commented on src/main/kotlin/racinggame/Car.kt:
```diff
@@ -1,25 +1,21 @@
 
-    fun move(diceResult: Int) {
-        position = CarCondition.getCondition(diceResult).apply(position)
+    init {
+        require(name.length in 1..5) {
+            "자동차 이름은 1~5자만 가능합니다."
+        }
```
이름에 대한 검증을 init 블록에서 처리 👍🏻

laco-dev commented on src/main/kotlin/racinggame/Main.kt:
```diff
@@ -1,16 +1,18 @@
     )
 
-    val result = game.play().translate()
+    val result = game.play()
+    val winners = game.getWinners()
```
승자의 결과는 result에서 만들어 지는 것이 자연스럽지 않을까요?
```kotlin
val result = game.play()
val winners = result.getWinners()
```

baek0318 replied:
Result라는 것이 단순히 데이터만을 전달하는 DTO 정도의 역할로만 생각을 해서 승자를 판별하는 로직이 포함되지 않는것이 더 좋다고 생각을 했습니다
그리고 승자를 판별하는것 또한 게임의 역할이라고 생각을 해서 승자 판별 기능을 game에서 호출하도록 만들었습니다
Result도 하나의 도메인으로 보는것이 적절할까요?

laco-dev commented on src/main/kotlin/racinggame/RacingGame.kt:
```diff
@@ -6,21 +6,30 @@ class RacingGame(
+        val num = dice.pitch()
+        car.move(num)
+        Car(car.name, car.position)
+    }
```
마이너한 의견입니다.
코드를 읽을 때에는 위에서 아래로 읽기 때문에 함수의 호출 순서를 위에서 아래로 맞춰주시면 좋습니다.

```kotlin
fun play() { 
   roundPlay(...)
}

private fun roundPlay()  {... }
```

laco-dev commented on src/main/kotlin/racinggame/RacingGame.kt:
```diff
@@ -6,21 +6,30 @@ class RacingGame(
-            car.move(num)
-            car.position
-        }
+        roundPlay(cars)
+    }.map { round ->
+        PlayResult(round)
+    }.let {
+        RacingGameResult(it)
```
아래는 취향의 영역 일 수는 있습니다.
함수형을 이용해서 체인 형태로 제공 하는 것이 좋을 때가 있지만, 함수가 어떻게 동작하는지 파악하는데에는 다소 어려움이 있을 수 있습니다.

저는 함수 내부적으로 상태를 변경하는 경우, 명확하게 stateless 방법으로 처리하고는 합니다.
```kotlin
fun play(): RacingGameResult {
    val racingResults = mutableListOf<PlayResult>()
    repeat(tryCount) {
        val racingResult = PlayResult(roundPlay(cars))
        racingResults.add(racingResult)
    }
    return RacingGameResult(racingResults)
}
```

baek0318 replied:
명확하게 stateless 방법으로 처리했다고 하셨는데 어떤 상태를 가지지 않았다는것인지 state하게 짜면 어떻게 되는지 궁금합니다

laco-dev replied:
아하 함수 내부에서의 프로세스에 대한 내용이였어요.
함수 내부에서 생성한 variable이 함수 밖으로 나가면서 사라지게 되니까요.

그래서 함수형으로 불변 map 형태로 이어나가는 대신 명확하게 variable의 상태를 바꾸는 전략을 의미했습니다.

laco-dev commented on src/main/kotlin/racinggame/RacingGame.kt:
```diff
@@ -6,21 +6,30 @@ class RacingGame(
-        }.reduce { s1, s2 ->
-            s1 + s2
-        }
+    fun getWinners() = cars.sortedByDescending { car ->
+        car.position
+    }.let { sortedCars ->
+        getJointWinner(sortedCars)
+    }.let {
+        PlayWinners(it)
+    }
```
비슷하게 코틀린의 let과 같은 범위 지정 함수는, 
단순히 변수로 안만들기 위해서 사용하는 것 보다는 적절한 사용 방법과 목적이 있습니다.
https://kotlinlang.org/docs/scope-functions.html

또한 함수가 복잡한 경우 반환 형이 무엇인지 알 수 있도록 명확하게 반환 타입을 작성 해 주시는 것이 좋습니다.
```kotlin
fun getWinners(): PlayWinners
```

baek0318 replied:
중괄호를 사용하는것을 최대한 지양하면서 작성하다보니 오히려 더 복잡한 코드가 만들어진듯 하네요
제가 용도를 정확하게 알지 못하고 사용한것 같아서 수정하는게 좋아보이네요

laco-dev commented on src/main/kotlin/racinggame/RacingGameResult.kt:
```diff
@@ -0,0 +1,35 @@
+    }
+}
+
+interface Result<T> {
+    fun translate(): T
+}
```
Result라는 인터페이스와 translate()는 왜 필요할까요?

어떠한 목적을 가지고 특정 타입으로 변환 가능한 Mapper 인터페이스 처럼 보이는데요.
저는 작성하신 의도의 목적이 UI에서 출력하기 위해 만들어진 것으로 보입니다.

하지만 도메인 모델 입장에서는 UI로 어떻게 출력할 지는 관심사가 아닙니다.
실제로 translate()는 출력 목적 외로는 사용이 되지 않고 가령 출력 방식이 달라질 경우 도메인 모델 수정이 불가피합니다.

baek0318 replied:
경우에 따라서 여러가지 Result가 생기게 되었고 공통적으로 Output에서 사용되기 위해서 거기에 맞는 형태의 데이터를 제공하기 위해서 translate라는 기능을 가지게 되었습니다
그러므로 Result라는 인터페이스를 만들게 되었습니다 위에서도 얘기를 했었지만 Result를 도메인으로 본다면 치명적?이라고 볼 수도 있겠지만 하나의 DTO로 본다면 크게 문제가 되나? 싶은 생각이 들었습니다
결론적으로는 XXResult라는 것을 단순 DTO로 볼지 아니면 하나의 도메인으로 볼지에 대해서 갈릴것 같다라는 생각이 드네요

laco-dev commented on src/main/kotlin/racinggame/RacingGameResult.kt:
```diff
@@ -0,0 +1,35 @@
+    }
+}
+
+data class PlayResult(
+    val result: List<Car>
+) : Result<List<String>> {
```
자동차 경주의 결과라고 한다면 PlayResult 대신 RoundResult가 의미로 적절하지 않을까요?
PlayResult라는 의미가 무엇에 대한 결과인지 예측이 다소 어려워 보입니다.

laco-dev commented on src/test/kotlin/racinggame/RacingGameTest.kt:
```diff
@@ -1,29 +1,74 @@
+                }
+            }
+        }
+        When("2번 시도의 4이상의 숫자가 나온 경우") {
```
바로 위의 테스트와 완전히 동일하지만, 시도 횟수가 증가함에 따라 불필요한 반복 코드가 많이 발생하게 되었는데요.

kotest의 파라미터 테스트 혹은 적절한 반복을 나누어서 효율적으로 작성 해 볼 수 있지 않을까요?
```kotlin
Given("4 이상의 주사위가 주어지면") {
   When("경주를 N바퀴 진행하면") {
       Then("모든 자동차들의 위치는 N이다") { ... }
   }
}
```

laco-dev reviewed (changes requested):
4단계 미션 잘 해주셨습니다.
불변으로 객체를 적절히 유지한 부분도 좋았습니다.

코틀린이 많은 편의 기능을 제공하지만 적절히 사용할 때 더욱 효과를 발휘합니다.
대표적으로 범위 지정함수가 그 예로 들 수 있을텐데요.
사용 예시들에 대해 찾아보시고 고민 해 보시면 좋을 것 같습니다.

baek0318 committed:
step4 : 명칭을 PlayResult -> RoundResult로 변경
- 한 개의 라운드의 결과값을 포함하기 때문에 Round로 변경

baek0318 committed:
step4 : private function의 위치 변경

baek0318 committed:
step4 : scope function을 상황에 맞는 형태로 변경
- 기존에 let을 사용했었는데 let보다는 정렬된 리스트에서 추가적으로 공동 우승자를 판별한다는 의미에서 also가 더 적절하다고 판단해서 수정

baek0318 committed:
step4 : RacingGameTest에서 중복된 테스트 코드를 작성하지 않도록 변경

laco-dev commented on src/main/kotlin/racinggame/RacingGame.kt:
```diff
@@ -7,20 +7,29 @@ class RacingGame(
-        }.reduce { s1, s2 ->
-            s1 + s2
-        }
+    fun getWinners() = cars.sortedByDescending { car ->
+        car.position
+    }.also { sortedCars ->
+        getJointWinner(sortedCars)
+    }.let {
+        PlayWinners(it)
+    }
```
> Result라는 것이 단순히 데이터만을 전달하는 DTO 정도의 역할로만 생각을 해서 승자를 판별하는 로직이 포함되지 않는것이 더 좋다고 생각을 했습니다
그리고 승자를 판별하는것 또한 게임의 역할이라고 생각을 해서 승자 판별 기능을 game에서 호출하도록 만들었습니다
Result도 하나의 도메인으로 보는것이 적절할까요?

그 의도라면, 플레이를 통해 결과를 만들어낼 때 승자를 같이 결정할 수 있지 않을까요?
현재는 경주와 우승자가 동시에 결정되는데, 별도의 함수로 각각 호출할 필요가 없습니다.

laco-dev commented on src/main/kotlin/racinggame/RacingGameResult.kt:
```diff
@@ -0,0 +1,35 @@
+    }
+}
+
+data class RoundResult(
+    val result: List<Car>
+) : Result<List<String>> {
+    override fun translate() = result.map { car ->
+        (0..car.position).map {
+            "-"
+        }.reduce { s1, s2 ->
+            s1 + s2
+        }.let {
+            "${car.name} : $it"
+        }
+    }
+}
```
> 경우에 따라서 여러가지 Result가 생기게 되었고 공통적으로 Output에서 사용되기 위해서 거기에 맞는 형태의 데이터를 제공하기 위해서 translate라는 기능을 가지게 되었습니다
그러므로 Result라는 인터페이스를 만들게 되었습니다 위에서도 얘기를 했었지만 Result를 도메인으로 본다면 치명적?이라고 볼 수도 있겠지만 하나의 DTO로 본다면 크게 문제가 되나? 싶은 생각이 들었습니다
결론적으로는 XXResult라는 것을 단순 DTO로 볼지 아니면 하나의 도메인으로 볼지에 대해서 갈릴것 같다라는 생각이 드네요

남겨주신 의도는 공감합니다.
하지만 DTO 라고 하더라도 명확하게 출력을 위한 처리를 담당하 고 있기 때문에 도메인 모델이라고는 볼 수 없습니다.
만약 DTO의 역할을 한다면 `RacingGameResult -> XXX -> Result` 형태로 중간 과정이 필요해 보입니다.

이 내용은 5단계 미션 요구사항을 같이 고려해보시면 도움이 될 것 같습니다.

laco-dev commented on src/test/kotlin/racinggame/RacingGameTest.kt:
```diff
@@ -1,28 +1,68 @@
     Given("자동차 게임 시작") {
-        When("4이상의 숫자가 나온 경우") {
-            val tryCount = 1
+        When("N번 시도의 4이상의 숫자가 나온 경우") {
```
kotest에서도 ParameterizedTest를 할 때 템플릿을 활용할 수 있어요.
나중에 기회가 되면 찾아보셔도 좋습니다.

laco-dev commented on src/test/kotlin/racinggame/RacingGameTest.kt:
```diff
@@ -1,28 +1,68 @@
+                            }
+                        }
+                    }
+                    if (tryCount == 2) {
+                        result.racingResult.forEachIndexed { index, racingResult ->
+                            if (index == 0) {
+                                racingResult.result.forEach { playResult ->
+                                    playResult.position shouldBe 1
+                                }
+                            }
+                            if (index == 1) {
+                                racingResult.result.forEach { playResult ->
+                                    playResult.position shouldBe 2
+                                }
+                            }
+                        }
+                    }
```
값을 가져오기 위해 여러 depth와 복잡한 로직이 필요하다면, 적절한 자료구조 혹은 함수를 통해 개선 할 수도 있어요.
가령 Round 에 해당하는 결과들을 가져오기 위해 List가 아닌 Map을 사용할 수 있습니다.

laco-dev reviewed (approved):
피드백 반영 잘 해주셨습니다.
고민하고 계신 부분은 다음 5단계 미션에서의 요구사항이기 때문에 참고해서 작업 해 주시면 될 것 같습니다.

질문주신 내용은 같이 남겨두었습니다.
